opencog/atomspace/AtomTable.cc

```diff
@@ -498,23 +495,45 @@ Handle AtomTable::add(AtomPtr atom, bool async)
     atom->keep_incoming_set();
     atom->setAtomSpace(_as);
 
+    ContentHash ch = atom->get_hash();
+
+    // Lock, to prevent two different threads from trying to add exactly
+    // the same atom.
+    std::unique_lock<std::recursive_mutex> lck(_store_mtx);
+
+    // Search again, now that we hold the lock. This is necessary for
+    // the case where another atom was just inserted by another
+    // thread while we were cloning and copying values in this thread.
+    auto range = _atom_store.equal_range(ch);
+    auto bkt = range.first;
+    auto end = range.second;
+    for (; bkt != end; bkt++) {
+        if (*((AtomPtr) bkt->second) == *atom) {
+            return bkt->second;
```

Comment:
well, this return is no longer legal, after you've built the incoming set for the atom.  Once you've built the incoming set, that's it, you can no longer abandon the transaction.  Or rather, if you did abandon, you would have to unbuild the incoming set here, before returning.  But that opens a race condition, if anyone happens to be traversing the incoming set at just that instant: they'll find atoms that are not in the atomspace

The core problem is that e.g. pattern matching, but also many other algos traverse the graph by following the incoming set of an atom. But this return means that there are now atoms in the incoming set, that are not in the atomsapce ... so the graph walk will walk out of the atomspace and start following junk -- viz atoms that are copies, and thus have bad TV's on them.  This will break the rule engine, which is driven by TV's.